src/plots/geo/geo2.js

```diff
@@ -0,0 +1,612 @@
+    this.hasChoropleth = false;
+    for(var i = 0; i < geoCalcData.length; i++) {
+        if(geoCalcData[i][0].trace.type === 'choropleth') {
+            this.hasChoropleth = true;
+            break;
+        }
+    }
+
+    this.updateProjection(fullLayout, geoLayout);
+    this.updateBaseLayers(fullLayout, geoLayout);
+    this.updateDims(fullLayout, geoLayout);
+
+    Plots.generalUpdatePerTraceModule(this, geoCalcData, geoLayout);
+
+    var scatterLayer = this.layers.frontplot.select('.scatterlayer');
+    this.dataPoints.point = scatterLayer.selectAll('.point');
+    this.dataPoints.text = scatterLayer.selectAll('text');
+    this.dataPaths.line = scatterLayer.selectAll('.js-line');
+
+    var choroplethLayer = this.layers.backplot.select('.choroplethlayer');
+    this.dataPaths.choropleth = choroplethLayer.selectAll('path');
+
+    this.updateFx(fullLayout, geoLayout);
+    this.render();
+};
+
+proto.updateProjection = function(fullLayout, geoLayout) {
+    var gs = fullLayout._size;
+    var domain = geoLayout.domain;
+    var projLayout = geoLayout.projection;
+    var rotation = projLayout.rotation || {};
+    var center = geoLayout.center || {};
+
+    var projection = this.projection = getProjection(geoLayout);
+
+    // set 'pre-fit' projection
+    projection
+        .center([center.lon - rotation.lon, center.lat - rotation.lat])
+        .rotate([-rotation.lon, -rotation.lat, rotation.roll])
+        .parallels(projLayout.parallels);
+
+    // setup subplot extent [[x0,y0], [x1,y1]]
+    var extent = [[
+        gs.l + gs.w * domain.x[0],
+        gs.t + gs.h * (1 - domain.y[1])
+    ], [
+        gs.l + gs.w * domain.x[1],
+        gs.t + gs.h * (1 - domain.y[0])
+    ]];
+
+    var lonaxis = geoLayout.lonaxis;
+    var lataxis = geoLayout.lataxis;
+    var rangeBox = makeRangeBox(lonaxis.range, lataxis.range);
+
+    // fit projection 'scale' and 'translate' to set lon/lat ranges
+    projection.fitExtent(extent, rangeBox);
+
+    var b = this.bounds = projection.getBounds(rangeBox);
+    var s = this.fitScale = projection.scale();
+    var t = projection.translate();
+
+    if(
+        !isFinite(b[0][0]) || !isFinite(b[0][1]) ||
+        !isFinite(b[1][0]) || !isFinite(b[1][1]) ||
+        isNaN(t[0]) || isNaN(t[0])
+    ) {
+        Lib.warn('Invalid geo settings');
+
+        // TODO fallback to default ???
+    }
+
+    // px coordinates of view mid-point,
+    // useful to update `geo.center` after interactions
+    var midPt = this.midPt = [
+        b[0][0] + (b[1][0] - b[0][0]) / 2,
+        b[0][1] + (b[1][1] - b[0][1]) / 2
+    ];
+
+    // adjust projection to user setting
+    projection
+        .scale(projLayout.scale * s)
+        .translate([t[0] + (midPt[0] - t[0]), t[1] + (midPt[1] - t[1])])
+        .clipExtent(b);
+
+    // the 'albers usa' projection does not expose a 'center' method
+    // so here's this hack to make it respond to 'geoLayout.center'
+    if(geoLayout._isAlbersUsa) {
```

Comment:
would there be a way to fold this into `getProjection`, perhaps where you fill in [`dummyFn`](https://github.com/plotly/plotly.js/pull/2030/commits/4c1ddb80dd39ae6d1ba5d51d21b95b36919a8bc5#diff-b4732bdd35eeb4830531de6212e422e2R501)? I guess that would have to also override `.translate` so perhaps it's more trouble than it's worth... still `_isAlbersUsa` is a bit annoying.

Reply:
There's no easy way unfortunately, if we add a special mocked `.center` method for `albersUsa`, we would also need to have special logic in `.fitExtent` which would otherwise clear the `translate` settings applied by `.center`.